### Que 1. What are the five key concepts of Object-Oriented Programming (OOP)?

The five key concepts of Object-Oriented Programming (OOP) are:

#### 1. Class: 
A blueprint for creating objects (instances). It defines a set of properties and methods that the objects created from the class will have.

#### 2. Object:
An instance of a class. Objects have attributes (data) and behavior (methods) defined by their class.

#### 3. Encapsulation: 
The practice of bundling data (attributes) and methods (functions) that operate on the data into a single unit (class), and restricting access to some of the object's components. This helps protect the integrity of the data by preventing external interference and misuse.

#### 4. Inheritance: 
A mechanism that allows one class (child or subclass) to inherit properties and methods from another class (parent or superclass). This promotes code reuse and logical hierarchy in class relationships.

#### 5. Polymorphism: 
The ability of different classes to respond to the same method call in different ways. It allows one interface to be used for a general class of actions, with the specific action determined by the type of object that implements the interface.

These principles form the foundation of OOP, enabling code modularity, reuse, and flexibility.

### Que 2. Write a Python class for a `Car` with attributes for `make`, `model`, and `year`. Include a method to display the car's information.


An example of a Python class for a Car with the attributes make, model, and year, and a method to display the car’s information:

In [23]:
class Car:
    def __init__(self, make, model, year):
        self.make = make
        self.model = model
        self.year = year

    def display_info(self):
        print(f"Car Information: {self.year} {self.make} {self.model}")

# Example usage:
my_car = Car("Hyundai", "Verna", 2022)
my_car.display_info()

Car Information: 2022 Hyundai Verna


#### In this code:
- The __init__ method is used to initialize the attributes make, model, and year when a new instance of the Car class is created.
- The display_info method prints the car’s information in a formatted string.

When you create an instance of the Car class (e.g., my_car = Car("Hyundai", "Verna", 2022)) and call my_car.display_info(), it will output the car's information.

### Que 3. Explain the difference between instance methods and class methods. Provide an example of each.

In Python, *instance methods* and *class methods* serve different purposes, though both are used within classes. Here's a breakdown of the differences:

#### 1. Instance Methods:
##### Definition: 
These methods are tied to an instance of the class and operate on the data (attributes) of a specific object (instance).
##### Access: 
They require an instance of the class to be called.
##### Self Parameter:
The first parameter is always self, which represents the instance calling the method. Through self, the method can access and modify the instance's attributes.
##### Use Case: 
Instance methods are used when you need to work with or modify an instance’s state.



#### Example of an Instance Method:

In [37]:
class Car:
       def __init__(self, brand, model):
           self.brand = brand
           self.model = model

       def get_info(self):
           return f"Car: {self.brand} {self.model}"

car1 = Car("Hyundai", "Creta") # Creating an instance of the class
print(car1.get_info())

Car: Hyundai Creta


#### 2. Class Methods:
##### Definition:
These methods are tied to the class itself rather than any instance. They can access or modify class-level data (shared across all instances).
##### Access:
They do not require an instance of the class to be called but can be called by both the class itself and its instances.
##### Cls Parameter: 
The first parameter is cls, which represents the class. Through cls, the method can access or modify class variables.
##### Use Case:
Class methods are useful when you want to perform actions that are related to the class but don't need instance-specific data.


#### Example of a Class Method:

In [40]:
class Car:
       car_count = 0  # Class variable

       def __init__(self, brand, model):
           self.brand = brand
           self.model = model
           Car.car_count += 1  # Incrementing class variable

       @classmethod
       def get_car_count(cls):
           return f"Total Cars: {cls.car_count}"

# Creating instances of the class
car1 = Car("Hyundai", "Creta")
car2 = Car("MG", "Hector")

# Accessing the class method
print(Car.get_car_count())

Total Cars: 2


#### Summary:
- Instance methods operate on specific instances of a class, using self to access instance data.
- Class methods operate on the class itself and use cls to access class-level data.

### Que 4. How does Python implement method overloading? Give an example.

In Python, *method overloading* (the ability to define multiple methods with the same name but different parameter signatures) is not supported in the same way as in languages like Java or C++. However, you can achieve similar behavior using default parameters or variable-length arguments (*args and **kwargs).

Python does not allow multiple methods with the same name in the same class. If you try to define multiple methods with the same name, only the last one will be used, as it will overwrite the previous ones.


#### 1. Use Default Arguments:
Default arguments allow you to define a single method that can be called with different numbers of parameters.

#### Example

In [65]:
class Calculator:
       def add(self, a, b=0, c=0):
           return a + b + c

calc = Calculator()

print(calc.add(20))        

20


In [66]:
print(calc.add(20, 6))

26


In [67]:
print(calc.add(20, 6, 4)) 

30


Here, the method add can handle one, two, or three arguments by providing default values for the additional parameters.


#### 2. **Use Variable-length Arguments (*args and *kwargs)**:
You can also use *args to pass a variable number of positional arguments or **kwargs for keyword arguments.

#### Example

In [69]:
class Calculator:
       def add(self, *args):
           return sum(args)

calc = Calculator()

print(calc.add(50))           

50


In [70]:
print(calc.add(50, 7))        

57


In [71]:
print(calc.add(50, 7, 3)) 

60


In this example, add can take any number of arguments, and all are summed.

#### 3. Using @singledispatch for Function Overloading:
While Python doesn't directly support method overloading, the functools.singledispatch decorator can be used for function overloading based on argument types. However, this works only for functions, not methods.


##### Example:

In [78]:
@singledispatch
def add(value):
       raise NotImplementedError("Unsupported type")

In [79]:
@add.register(int)
def _(value):
       return value + 1

In [80]:
print(add(10))

11


In [83]:
@add.register(str)
def _(value):
       return value + " is a word"

In [84]:
print(add("Hello"))

Hello is a word


Here, the add function is overloaded based on the argument type.

#### Summary:
- Python doesn't support traditional method overloading like some other languages.
- You can use default parameters or variable-length arguments (*args and **kwargs) to simulate overloading.
- @singledispatch allows function overloading based on argument type but is limited to functions, not methods.

### Que 5. What are the three types of access modifiers in Python? How are they denoted?

In Python, access modifiers control the visibility of class members (attributes and methods) from outside the class. Unlike languages such as Java or C++, Python doesn't have strict access control. Instead, it uses naming conventions to indicate different levels of access. Python's philosophy is based on the idea of "we are all consenting adults here," meaning it relies more on conventions than enforcement.

#### 1. Public Members:
##### Definition:
Public members are accessible from anywhere (both inside and outside the class).
##### Denotation: 
Public members are denoted by names that do not start with an underscore.
##### Usage: 
You can access these members freely from any part of the program.

##### Example:

In [89]:
class MyClass:
       def __init__(self):
           self.public_var = "I am public"
    

obj = MyClass()
print(obj.public_var)          

I am public


In [87]:
def public_method(self):
           return "This is a public method"

In [88]:
print(obj.public_method()) 

This is a public method


#### 2. *Protected Members*
##### Definition: 
Protected members are intended to be accessed only within the class and its subclasses. However, in Python, they can still be accessed from outside the class—this is merely a convention, not enforcement.
##### Denotation:
Protected members are denoted by a single leading underscore (_).
##### Usage:
By convention, protected members should not be accessed directly outside the class or its subclasses, although Python doesn’t strictly prevent it.

##### Example:

In [100]:
class MyClass:
       def __init__(self):
           self._protected_var = "I am happy"   

obj = MyClass()
print(obj._protected_var)        

I am happy


In [101]:
class MyClass:
       def _protected_method(self):
           return "This is a protected method"
obj = MyClass()
print(obj._protected_method())

This is a protected method


#### 3. Private Members:
##### Definition:
Private members are meant to be accessed only within the class that defines them. Python uses name mangling to make private members less accessible from outside the class.
##### Denotation:
Private members are denoted by a double leading underscore (__).
##### Usage:
Private members are not directly accessible outside the class. Python performs name mangling by prefixing the class name to the private member’s name, making it harder (but not impossible) to access them from outside the class.

##### Example:

In [145]:
class MyClass:
       def __init__(self):
           self.__private_var = "I am private"
       
       def __private_method(self):
           return "This is a private method"

obj = MyClass()

# These will raise an AttributeError if accessed directly
# print(obj.__private_var)  # AttributeError: 'MyClass' object has no attribute '__private_var'
# print(obj.__private_method())  # AttributeError: 'MyClass' object has no attribute '__private_method'

# But can be accessed through name mangling
print(obj._MyClass__private_var)        
print(obj._MyClass__private_method())    

I am private
This is a private method


#### Summary:
##### Public members:
No leading underscore, accessible from anywhere.
##### Protected members:
A single leading underscore (_), intended for internal use within the class and subclasses.
##### Private members:
A double leading underscore (__), intended to be fully private, but can still be accessed through name mangling (_ClassName__member).

### Que 6. Describe the five types of inheritance in Python. Provide a simple example of multiple inheritance.

In Python, inheritance allows one class (child class or subclass) to inherit attributes and methods from another class (parent class or superclass). There are five types of inheritance in Python, each defining a different way in which classes can be related to one another.

#### 1. Single Inheritance:
A subclass inherits from only one parent class.

##### Example:

In [161]:
class Animal:
         def sound(self):
             return "Some sound"

class Dog(Animal):
         def bark(self):
             return "Woof!"

dog = Dog()
print(dog.sound())  
print(dog.bark())  

Some sound
Woof!


#### 2. Multiple Inheritance
A subclass can inherit from more than one parent class.
##### Example:


In [169]:
class Father:
         def skills(self):
             return "Driving"

class Mother:
         def skills(self):
             return "Cooking"

class Child(Father, Mother):
         pass

child = Child()
print(child.skills())  # Output: Driving (method resolution order chooses Father first)
     

Driving


In this example, the child class inherits from both Father and Mother.

#### 3. Multilevel Inheritance:
A subclass inherits from a parent class, and another class inherits from that subclass, forming a chain

##### Example:

In [186]:
class Animal:
         def sound(self):
             return "Some sound"

class Dog(Animal):
         def bark(self):
             return "Woof!"

class Bulldog(Dog):
         def breed(self):
             return "Bulldog"

bulldog = Bulldog()
print(bulldog.sound())  # Output: Some sound
print(bulldog.bark())   # Output: Woof!
print(bulldog.breed())  # Output: Bulldog
     

Some sound
Woof!
Bulldog


#### 4. Hierarchical Inheritance:
Multiple subclasses inherit from the same parent class.
##### Example:

In [187]:
class Animal:
         def sound(self):
             return "Some sound"

class Dog(Animal):
         def bark(self):
             return "Woof!"

class Cat(Animal):
         def meow(self):
             return "Meow!"

dog = Dog()
cat = Cat()
print(dog.sound())  
print(cat.sound()) 

Some sound
Some sound


#### 5. Hybrid Inheritance
A combination of two or more types of inheritance (e.g., a mix of multiple and hierarchical inheritance).


##### Example:

In [190]:
 class Animal:
    def sound(self):
             return "Some sound"

class Dog(Animal):
    def bark(self):
             return "Woof!"

class Cat(Animal):
    def meow(self):
             return "Meow!"

class Pet(Dog, Cat):
    def activity(self):
             return "Plays with humans"

pet = Pet()
print(pet.sound())    
print(pet.bark())     
print(pet.meow()) 

Some sound
Woof!
Meow!


In [191]:
print(pet.activity())

Plays with humans


#### Example of Multiple Inheritance:

In [194]:
class Father:
    def profession(self):
        return "Engineer"

class Mother:
    def hobby(self):
        return "Painting"

class Child(Father, Mother):
    def activity(self):
        return "Plays video games"

# Create an instance of Child class
child = Child()
# Access methods from both parent classes
print(child.profession())  
print(child.hobby())       
print(child.activity())

Engineer
Painting
Plays video games


#### Summary:
##### 1. Single Inheritance: 
A class inherits from one parent class.
##### 2. Multiple Inheritance: 
A class inherits from more than one parent class.
##### 3. Multilevel Inheritance:
A class inherits from a parent class, which is itself inherited from another class.
##### 4. Hierarchical Inheritance:
Multiple classes inherit from the same parent class.
##### 5. Hybrid Inheritance:
A combination of two or more types of inheritance.

### Que 7. What is the Method Resolution Order (MRO) in Python? How can you retrieve it programmatically?

#### Method Resolution Order (MRO) in Python:

The Method Resolution Order (MRO) in Python determines the order in which classes are searched when invoking a method or accessing an attribute in the presence of inheritance, especially in multiple inheritance scenarios. When you call a method on an instance, Python follows the MRO to find where that method is defined.

Python uses the *C3 linearization* algorithm (also called C3 superclass linearization) to create the MRO in a way that maintains a consistent method search path. This ensures that classes are searched in a predictable order, and ambiguities are resolved in multiple inheritance situations.

#### Key Points:
- The MRO determines the sequence in which base classes are checked when a method is called.
- In single inheritance, the MRO is straightforward: the class itself and its direct parent classes are checked in sequence.
- In multiple inheritance, Python uses the C3 linearization algorithm to ensure a consistent and unambiguous resolution order.

#### Retrieving MRO Programmatically:

There are two primary ways to retrieve the MRO of a class:

#### 1. Using the __mro__ attribute:
Every class in Python has an attribute called __mro__, which stores the MRO as a tuple of classes.

##### Example:

In [199]:
class A:
         pass

class B(A):
         pass

class C(B):
         pass

print(C.__mro__)

(<class '__main__.C'>, <class '__main__.B'>, <class '__main__.A'>, <class 'object'>)


#### 2. Using the mro() method:
You can also call the mro() method on a class to retrieve the MRO as a list.

##### Example:

In [200]:
class A:
         pass

class B(A):
         pass

class C(B):
         pass

print(C.mro())

[<class '__main__.C'>, <class '__main__.B'>, <class '__main__.A'>, <class 'object'>]


#### Example with Multiple Inheritance

In [201]:
class A:
    def who_am_i(self):
        return "I am A"

class B(A):
    def who_am_i(self):
        return "I am B"

class C(A):
    def who_am_i(self):
        return "I am C"

class D(B, C):
    pass

d = D()
print(d.who_am_i()) 
print(D.mro()) 

I am B
[<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>]


In this example:
- Class D inherits from both B and C.
- The method who_am_i() is defined in both B and C, but Python will use the method from B because B comes before C in the MRO.
- The MRO for class D is [D, B, C, A, object].

#### How MRO Works:
- For single inheritance, the MRO is simple: it follows the class hierarchy from the current class to its parent classes, and ultimately to the built-in object class.
- For multiple inheritance, Python uses the C3 linearization algorithm to determine the order in which classes are considered. It ensures:
 ##### 1. Child class first: 
 The class itself is checked first.
 ##### 2. Depth-first, left-to-right: 
 Python checks base classes in a depth-first, left-to-right order, meaning it searches the base classes in the order they are inherited.
 ##### 3. No class is repeated:
 Each class appears in the MRO only once, even if it appears multiple times in the inheritance hierarchy.

#### Summary:
- **MRO** determines the order in which base classes are searched for a method or attribute.
- **C3 linearization** is the algorithm Python uses to ensure a consistent and unambiguous search order.
- You can retrieve the MRO using the __mro__ attribute or the mro() method.

### Que 8. Create an abstract base class `Shape` with an abstract method `area()`. Then create two subclasses`Circle` and `Rectangle` that implement the `area()` method.

A Python implementation of an abstract base class Shape with an abstract method area(), and two subclasses Circle and Rectangle that implement the area() method:

In [202]:
from abc import ABC, abstractmethod
import math

In [205]:
# Abstract base class
class Shape(ABC):
    
    @abstractmethod
    def area(self):
        pass

# Subclass for Circle
class Circle(Shape):
    
    def __init__(self, radius):
        self.radius = radius
    
    def area(self):
        return math.pi * self.radius ** 2

# Subclass for Rectangle
class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height

##### Example usage:

In [206]:
circle = Circle(5)
rectangle = Rectangle(4, 6)

print("Circle area:", circle.area())        
print("Rectangle area:", rectangle.area())

Circle area: 78.53981633974483
Rectangle area: 24


#### Explanation:
##### 1. Shape:
This is the abstract base class that contains the abstract method area(). It serves as a template for its subclasses.
##### 2. Circle: 
This subclass implements the area() method to compute the area of a circle using the formula π * r^2.
##### 3. Rectangle: 
This subclass implements the area() method to compute the area of a rectangle using the formula width * height.

Each subclass must implement the area() method; otherwise, it will raise an error if instantiated.

### Que 9. Demonstrate polymorphism by creating a function that can work with different shape objects to calculate and print their areas.

An example of how you can demonstrate polymorphism by creating a function that works with different Shape objects to calculate and print their areas:

##### Function to calculate and print area of any shape

In [207]:
def print_area(shape):
    print(f"The area of the {shape.__class__.__name__} is: {shape.area()}")

##### Example usage:

In [209]:
circle = Circle(5)
rectangle = Rectangle(4, 6)

# Polymorphism in action
print_area(circle)      
print_area(rectangle) 

The area of the Circle is: 78.53981633974483
The area of the Rectangle is: 24


#### Explanation:
- The function print_area() takes an object of type Shape (or any subclass of Shape) and calls its area() method.
- This is an example of *polymorphism*, where the same function (print_area) behaves differently based on the type of the object passed to it.
- The function is able to work with any object that is a subclass of Shape without needing to know its specific type (Circle or Rectangle). The appropriate area() method is called for each type.

#### Output:
plaintext
- The area of the Circle is: 78.53981633974483
- The area of the Rectangle is: 24

### Que 10: Implement encapsulation in a `BankAccount` class with private attributes for `balance` and `account_number`. Include methods for deposit, withdrawal, and balance inquiry.

Implementing encapsulation in a BankAccount class with private attributes and methods for deposit, withdrawal, and balance inquiry:

In [2]:
class BankAccount:
    def __init__(self, account_number, initial_balance=0):
        # Private attributes
        self.__account_number = account_number
        self.__balance = initial_balance

    # Method to deposit money
    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            print(f"Deposited {amount}. New balance: {self.__balance}")
        else:
            print("Deposit amount must be positive.")

    # Method to withdraw money
    def withdraw(self, amount):
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"Withdrew {amount}. New balance: {self.__balance}")
        else:
            print("Invalid withdrawal amount.")

    # Method to inquire the balance
    def get_balance(self):
        return self.__balance

    # Method to get the account number
    def get_account_number(self):
        return self.__account_number

# Example usage
account = BankAccount("987654321", 500)
account.deposit(200)            
account.withdraw(100)           
print(account.get_balance())    
print(account.get_account_number()) 

Deposited 200. New balance: 700
Withdrew 100. New balance: 600
600
987654321


#### Key Concepts:


-Private Attributes: The __account_number and __balance are private attributes, which means they cannot be accessed directly from outside the class. They can only be accessed and modified through the class methods.


-Encapsulation: The private attributes ensure that the internal state of the object is hidden from the outside, and any interaction with the data is controlled through methods.


-Public Methods: The methods deposit, withdraw, get_balance, and get_account_number are public methods, allowing controlled access and manipulation of the private data.


-This structure ensures that the integrity of the BankAccount class is maintained, with all operations on the account balance being performed through the appropriate methods.

### Que 11: Write a class that overrides the `__str__` and `__add__` magic methods. What will these methods allow you to do?

Certainly! When you override the __str__ and __add__ magic methods in a Python, you enable custom behaviors for converting objects to strings and for adding objects together with the + operator.

Here’s an example of a class that overrides both of these methods:

In [6]:
class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __str__(self):
        return f"{self.real} + {self.imaginary}i"

    def __add__(self, other):
        if isinstance(other, ComplexNumber):
            new_real = self.real + other.real
            new_imaginary = self.imaginary + other.imaginary
            return ComplexNumber(new_real, new_imaginary)
        return NotImplemented

# Example usage
c1 = ComplexNumber(3, 4)
c2 = ComplexNumber(1, 2)

print(c1)  
print(c2)  

c3 = c1 + c2  
print(c3)  

3 + 4i
1 + 2i
4 + 6i


#### Explanation:


##### __str__ Method:


The __str__ method is used to define the string representation of an object.

    
When you use print() on an object or call str() on it, Python uses the __str__ method to determine what to display.


In this example, __str__ is overridden to provide a formatted string representation of a complex number, e.g., "3 + 4i".


__add__ Method:

The __add__ method defines how two objects of the class should be added together using the + operator.


In the example, __add__ is overridden to add two ComplexNumber objects by separately adding their real and imaginary parts.

    
When you write c1 + c2, Python automatically calls the __add__ method, returning a new ComplexNumber object that represents the sum of the two.

    
What These Methods Allow You to Do:


Custom String Representation: By overriding __str__, you can control how your objects are represented as strings, making it easier to display and understand the object's data when printed or logged.



Operator Overloading: By overriding __add__, you can define how the + operator works for your objects, enabling intuitive mathematical operations that align with the nature of the objects (e.g., adding complex numbers, points, etc.).



These techniques allow your class to behave more naturally and integrate smoothly with Python’s built-in functions and operators.

#### Key Concepts:


Custom String Representation: The __str__ method allows objects to be easily converted to a human-readable string, which is useful for debugging or logging.


Operator Overloading: The __add__ method (and similar magic methods like __sub__, __mul__, etc.) allows you to define custom behavior for operators, making objects behave in intuitive ways when combined with standard operators.

### Que 12: Create a decorator that measures and prints the execution time of a function.

Creating a decorator in Python that measures and prints the execution time of a function using the time module. Here's how you can do it:

In [7]:
import time

def time_it(func):
    def wrapper(*args, **kwargs):
        start_time = time.time()  # Record the start time
        result = func(*args, **kwargs)  # Call the original function
        end_time = time.time()  # Record the end time
        execution_time = end_time - start_time  # Calculate the execution time
        print(f"Execution time of {func.__name__}: {execution_time:.6f} seconds")
        return result  # Return the result of the original function
    return wrapper

# Example usage
@time_it
def example_function(n):
    total = 0
    for i in range(n):
        total += i
    return total

# Call the function
result = example_function(1000000)
print("Result:", result)

Execution time of example_function: 0.166045 seconds
Result: 499999500000


##### Explanation:


Decorator time_it:

1.The decorator time_it takes a function func as an argument and returns a wrapper function.

    
2.Inside wrapper, the start time is recorded using time.time().

    
3.The original function is then executed with any arguments passed to it.

    
4.After the function call, the end time is recorded.


5.The execution time is calculated by subtracting the start time from the end time.

    
6.The execution time is printed with the function's name.


7.Finally, the result of the original function is returned.


Applying the Decorator:

You apply the time_it decorator to a function using the @time_it syntax. This causes the decorator to wrap the original function with the wrapper function, adding the timing logic.

    
Example Function:

The example_function computes the sum of numbers from 0 to n-1.


When you call example_function(1000000), the execution time is measured and printed.

##### Output:

When you run the code, you'll see the execution time printed, along with the result of the function:

In [8]:
#Execution time of example_function: 0.030000 seconds
#Result: 499999500000

This decorator can be reused to measure the execution time of any function by simply applying @time_it to it.

### Que 13:  Explain the concept of the Diamond Problem in multiple inheritance. How does Python resolve it?

The Diamond Problem is a common issue in object-oriented programming that occurs in languages that support multiple inheritance. It arises when a class inherits from two classes that both inherit from a single common base class, forming a diamond-shaped inheritance structure. The problem is related to ambiguity about which method or attribute should be inherited from the base class.

The Diamond Problem
Consider the following class hierarchy:

      A
     / \
    B   C
     \ /
      D

Class A is the base class.

Classes B and C both inherit from class A.

Class D inherits from both B and C.

Here’s the diamond structure:


In [10]:
class A:
    def show(self):
        print("A")

class B(A):
    def show(self):
        print("B")

class C(A):
    def show(self):
        print("C")

class D(B, C):
    pass

Now, if you create an instance of D and call the show() method:

In [11]:
d = D()
d.show() 

B


The question arises: which show() method should be called—B's, C's, or A's? This ambiguity is the essence of the Diamond Problem.

How Python Resolves the Diamond Problem-

Python resolves the Diamond Problem using a method resolution order (MRO) and the C3 linearization algorithm. The MRO is the order in which base classes are searched when executing a method. This order ensures that a method is called from the correct class and avoids ambiguity.

In the example above, Python will search for methods in the following order (MRO):

D

B

C

A

This MRO can be observed using the __mro__ attribute or the mro() method:

In [12]:
print(D.mro())

[<class '__main__.D'>, <class '__main__.B'>, <class '__main__.C'>, <class '__main__.A'>, <class 'object'>]


So, in the **example**:

When d.show() is called, Python looks first in D for a show method.

Since D doesn’t have a show method, it proceeds to B and finds the show method there.

The show method from B is executed, so the output is B.

Why This Works

Python's MRO ensures that:

Each base class is only called once, even if it's inherited through multiple paths (this avoids duplicate calls to A).

The order is consistent and predictable, preventing ambiguity.
    
If class D had its own show method, it would be used, but if not, Python follows the MRO to find the next method.

Summary

The Diamond Problem occurs in multiple inheritance scenarios when ambiguity arises about which method or attribute to inherit from a common base class.

Python resolves the Diamond Problem using the MRO, which follows a linearization order based on the C3 algorithm.
    
The MRO ensures a consistent and unambiguous order of method resolution, preventing issues like duplicate method calls or ambiguity.








### Que 14: Write a class method that keeps track of the number of instances created from a class. 

You can keep track of the number of instances created from a class by using a class-level attribute that is incremented each time a new instance is created. Here's how you can implement this:

In [19]:
class InstanceCounter:
    instance_count = 0  # Class-level attribute to keep track of instances

    def __init__(self):
        InstanceCounter.instance_count += 1  # Increment count each time an instance is created

    @classmethod
    def get_instance_count(cls):
        return cls.instance_count  # Return the current count of instances

# Example usage
a = InstanceCounter()
b = InstanceCounter()
c = InstanceCounter()

print(InstanceCounter.get_instance_count())  


3


##### Explanation:

instance_count:

This is a class-level attribute (also known as a class variable). It is shared by all instances of the class and is used to keep track of how many instances of the class have been created.
    
It is initialized to 0.
    
__init__ Method:

Each time a new instance is created, the __init__ method is called. Inside this method, InstanceCounter.instance_count is incremented by 1.
    
This ensures that each time a new instance is created, the count is updated.
    
@classmethod and get_instance_count:

get_instance_count is a class method that returns the current value of instance_count.

The @classmethod decorator allows the method to be called on the class itself (cls) rather than on an instance of the class.
    
This method can be called on the class to check how many instances have been created so far.

##### Output:

If you create three instances of the InstanceCounter class, as shown in the example:

python-

Copy code
    
a = InstanceCounter()
    
b = InstanceCounter()
 
c = InstanceCounter()
 
Then calling InstanceCounter.get_instance_count() will return 3, indicating that three instances have been created.

This pattern is useful when you need to monitor or limit the number of instances that can be created from a class.

### Que 15:  Implement a static method in a class that checks if a given year is a leap year.

You can implement a static method to check if a given year is a leap year as follows:

In [14]:
class DateUtils:
    @staticmethod
    def is_leap_year(year):
        # A year is a leap year if it is divisible by 4
        # but not divisible by 100, except if it is divisible by 400
        if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
            return True
        else:
            return False

# Example usage
print(DateUtils.is_leap_year(2020))  
print(DateUtils.is_leap_year(1900)) 
print(DateUtils.is_leap_year(2000)) 
print(DateUtils.is_leap_year(2023)) 

True
False
True
False


#### Explanation:

@staticmethod Decorator:

The @staticmethod decorator indicates that the method does not depend on the instance of the class (i.e., it doesn’t access or modify the instance or class-level attributes). It operates solely based on the input arguments.

This makes is_leap_year a method that can be called directly on the class without creating an instance.

is_leap_year Method:

The logic inside the is_leap_year method follows the rules for determining leap years:
                                                          
A year is a leap year if it is divisible by 4.
    
However, if the year is divisible by 100, it is not a leap year unless it is also divisible by 400.
    
The method returns True if the year is a leap year and False otherwise.
    
Usage:
    
You can call this static method directly on the class without needing to create an instance:

In [17]:
print(DateUtils.is_leap_year(2020))

True


This method correctly identifies leap years according to the rules of the Gregorian calendar.